src/clap/fmt_span.rs

```diff
@@ -0,0 +1,62 @@
+//! Adapter for pasing the [`FmtSpan`] type.
+
+use clap::builder::EnumValueParser;
+use clap::builder::PossibleValue;
+use clap::builder::TypedValueParser;
+use clap::builder::ValueParserFactory;
+use tracing_subscriber::fmt::format::FmtSpan;
+
+/// Wrapper around [`FmtSpan`].
+#[derive(Clone)]
+pub struct FmtSpanWrapper(FmtSpan);
+
+impl From<FmtSpanWrapper> for FmtSpan {
+    fn from(value: FmtSpanWrapper) -> Self {
+        value.0
+    }
+}
+
+impl clap::ValueEnum for FmtSpanWrapper {
+    fn value_variants<'a>() -> &'a [Self] {
+        &[
+            Self(FmtSpan::NEW),
+            Self(FmtSpan::ENTER),
+            Self(FmtSpan::EXIT),
+            Self(FmtSpan::CLOSE),
+            Self(FmtSpan::NONE),
+            Self(FmtSpan::ACTIVE),
+            Self(FmtSpan::FULL),
+        ]
+    }
```

Comment:
1. I'm surprised this lifetime can't be elided?

2. Could this list be generated with the Rust equivalent of `[minBound .. maxBound]` in Haskell? The compiler won't tell you if a new variant is added, making this list incomplete.

Reply:
1. Me too! I guess the trait definition has an explicit lifetime so you're not allowed to elide it here?

<details><summary>Compilation error when I elide the lifetime</summary>


```
   error[E0106]: missing lifetime specifier
  --> src/clap/fmt_span.rs:20:28
   |
20 |     fn value_variants() -> &[Self] {
   |                            ^ expected named lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
help: consider using the `'static` lifetime
   |
20 |     fn value_variants() -> &'static [Self] {
   |                             +++++++

error[E0195]: lifetime parameters or bounds on method `value_variants` do not match the trait declaration
  --> src/clap/fmt_span.rs:20:22
   |
20 |     fn value_variants() -> &[Self] {
   |                      ^ lifetimes do not match method in trait

Some errors have detailed explanations: E0106, E0195.
For more information about an error, try `rustc --explain E0106`.
```

</details>

2. I *wish!* This has been [an open issue since before Rust 1.0](https://github.com/rust-lang/rfcs/issues/284#issuecomment-1592343574) (peep the Ruby-like syntax in the [original issue here](https://github.com/rust-lang/rust/issues/5417)). There's a few crates that implement this functionality for `enum` types with `#[derive]` macros, but they're pretty heavy so we don't use them here.